README.md
# himawari

![tech-debt](https://img.shields.io/codeclimate/tech-debt/engura/himawari)
[![maintainability](https://img.shields.io/codeclimate/maintainability/engura/himawari)](https://codeclimate.com/github/engura/himawari)
[![codefactor](https://img.shields.io/codefactor/grade/github/engura/himawari/main)](https://www.codefactor.io/repository/github/engura/himawari)
[![Inline docs](https://inch-ci.org/github/engura/himawari.svg?branch=main)](https://inch-ci.org/github/engura/himawari)
[![build](https://img.shields.io/circleci/build/github/engura/himawari)](https://app.circleci.com/pipelines/github/engura/himawari)

[![gem_version](https://img.shields.io/gem/v/himawari)](https://rubygems.org/gems/himawari)
[![gem_dl](https://img.shields.io/gem/dt/himawari)](https://rubygems.org/gems/himawari)
![lines](https://img.shields.io/tokei/lines/github/engura/himawari)
![size](https://img.shields.io/github/languages/code-size/engura/himawari)
[![license](https://img.shields.io/github/license/engura/himawari)](https://opensource.org/licenses/MIT)

<details>
 <summary>TOC</summary>

- [about](#himawari)
- [installing](#installation)
- [using](#usage)
- [todo](#todo)
- [development](#development)
- [contributing](#contributing)
- [license](#license)
</details>

Access images from the [himawari8 weather satellite](https://himawari8.nict.go.jp)] (courtesy of NICT) and compose near real-time desktop backgrounds of Earth or use the high-res images for other personal uses. For example....
![full globe](doc/img/h_2020-06-01T0100.png)
![high res section](doc/img/h_2020-11-29T0110.png)

# Installation
Add this line to your application's Gemfile:
```
gem 'himawari'
```
And then execute:
```
bundle
```
Or install it yourself as:
```
gem install himawari
```
## Install CLI Dependencies
Mac/Linux: run `bin/setup`
On other systems, please try to install `imagemagick` and `parallel` utilities manually...
https://linux.die.net/man/1/parallel
https://github.com/minimagick/minimagick

# Usage

You can use it as a stand alone CLI executable, or as a Library.

*RUN EXAMPLE*: `himawari -w /home/User/himawari -d /home/User/Pictures/live -r 4 -f top -m day -v`

To set cron (with all the supplied arguments):
`himawari -w /home/User/himawari -d /home/User/Pictures/live -r 4 -f top -m day -v -c set`
and then to remove it from cron (please note the arguments must be the same as the ones used during `cron set`-ting, otherwise it won't be removed):
`himawari -w /home/User/himawari -d /home/User/Pictures/live -r 4 -f top -m day -v -c clear`

### Details
 - Grabs a tiled image from Himawari, (each photo of Earth is split into square tiles of 550x550px) reassembles it into one image, and optionally copies one of the downloaded images into a `destination` folder to use as a desktop background.
![a tile](doc/img/t_2020-12-01T0410-1_0.png)
 - Can change the resolution. Default is "2", smallest. Maximum is "20". It will produce an jpg of ~200MB in size for the full planet. The allowed steps are [2, 4, 8, 16, 20]. Resolution of 2 means that the image is composed of *2* tiles across, so the full image becomes 1100px wide/high. An image of resolution *factor 4* is then 550 * 4 = 2200px across. And the maximum one is 20 * 550 = 11000px X 11000px!!!!!
 - Since the images can be kinda big, we can customize which part of the planet we want to see: `top`, `full` planet, or bottom (`low`).
 - When using the `autorun` option (it's the method that command line utility uses), we save the last 48hrs in a folder (`workdir`). We can then have 2 options: either show what is `now` outside, or! cycle the last complete 24 hours with a new background photo every 2 minutes? Because the images are all downloaded at once, and then an incremental download of one photo as needed, the internet traffic is relatively reasonable. Cycling the background photos doesn't need to access internet or download anything.
 - An easy way to use the script: set the `himawari` into cron with whatever arguments you want it to use, and it will automatically update the photos in the destination folder every 2 minutes. Then, just use the OS's Desktop Settings to do a "desktop slideshow" using pictures in the `destination` folder.
 - Another way to use it can be: download whichever images you like into the `workdir`, and set the System's Desktop Settings to do a slideshow desktop background from there. The tricky part here is that the OS does not always pick up new images that are downloaded into the `workdir` right away. (maybe until a restart or something.) Also, the slideshows are randomized by default, so the picture of the planet could be jumping to random time stamps.

 - 1. do NOT attempt to fetch data from black listed WiFis
 - 2. if no internet, just show last available data
 - 3. keep only 1 day's worth of images (~144 photos)
 - 4. check for `No Image` images and skip those (until? if? --so far, I haven't noticed that they ever get changed) they become available. (There are tricky tiles that are literally just black squares with the words "No Image" written on them. They probably happen due to the Sun being in the camera's view. Or maybe because of Aliens. Or evil Government cover ups? Either way, these tiles are not useful to us, and are skipped.)
 ![no image](lib/himawari/no_image.png)

### Full list of arguments for the command line
`himawari -h`
- `-f, --focus STRING` Which section of the planet to focus on? Valid values are `full`, `top`, `mid`, `low`. Default is `top`. The script takes a common monitor's aspect ratio (16x9) and does its best to keep the cropped image as close to that AR as possible. It will not, however, crop off parts of the downloaded tiles, so the final image usually ends up with a wider (more squarish) aspect ratio than 16x9. (But never narrower.)
- `-m, --mode STRING` Valid values are `day` (cycles pics to the `destination` folder from the most recent day) or `live` (copies the latest photo downloaded to `destination`) Default is `day`. Right now, it is hard-coded to switch the photo each 2 minutes in the `destination` folder. With 144 photos to choose from in a 24 hour period, this ends up cycling the view of the planet 5 times each day. 
- `-r, --resolution INT` Adjust the resolution factor of the downloaded image. Valid numbers are 2, 4, 8, 16, 20. 20 is the highest resolution and 2 is the default. For a 4k-monitor a setting of 4 seems sufficient.
- `-d, --destination PATH` The folder where to copy a background image. If left blank, images will just be downloaded to `workdir`.
- `-w, --workdir PATH` The folder where to save all the downloaded pics. If left blank, images will be saved to the `./data` directory relative to your current path.(pwd)
- `-b, --blacklist STRING,STRING...` Blacklist SSIDs of networks from which we do not want to go online to download new images.
- `-c, --cron STRING` Can `set`/`clear` cron with the specified params, so we can update/download the images automatically
- `-v, --verbose` Increase verbosity: mostly for debugging
- `-s, --schedule` Flag for determining when the script is run by schedule/automatically. Don't worry about this one, it is set automatically when the script is added to cron.

### Usage as a library
```
require 'himawari'

# there are basically 2 useful methods exposed to play around with.
# To get 1 picture closest to the provided timestamp:
Himawari.get_pic(datetime: timestamp, workdir: '/home/user/himawari_pics')

# And to get a bunch of pics between 2 date-times:
params = { from: timestamp1, to: timestamp2, workdir: workdir }
Himawari.get_pics(params)
```
- `timestamp` can be either a string (like '2020-06-01 12:30') or a `Time` object.
- pretty much all of the `params` are optional and will be filled in with some defaults (same as the ones mentioned above) if left blank.
- the timestamps' defaults are set like this: `round_down_to_nearest_10_minute_mark(now - 10.minutes)`. So, if it is `2020-12-01 14:19` right now, default timestamp would be `2020-12-01 14:00`)
- these methods return `true` on success, `false` on failure.
- Feel free to check out some other usages in the `./specs`

# TODO
- improve documentation
- add [Windows support](https://github.com/engura/himawari/issues/1)?

# Development
After checking out the repo, doing the steps in `installation` above and messing around with the code, run `rake test` and `rubocop` to use the tests and make sure everything is ok. To run a specific test, use `rake test TEST=spec/test_base.rb TESTOPTS="--name=test_bad_params --seed=1234"` and as for rubocop: `rubocop lib/himawari/base.rb`

## Building the gem on local machine/from source
navigate to the gem's directory and...
Manually:
```
gem build himawari.gemspec
gem install himawari-#.#.#.gem # replace the # with the most recent version
```
Semi-manually (the end result is same as above):
```
bundle exec rake install
```

## Releasing a new version of the Gem
 - Update the version number in `himawari.gemspec`
 - run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the .gem file to [rubygems.org](https://rubygems.org).

# Contributing
Bug reports and pull requests are welcome on GitHub at https://github.com/engura/himawari.

# License
The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# 0.1.2
- Updated Readme with more details
- Updated meta links in gemspec
- Fixed a crash/bug related to missing a png file (I forgot that we actually DO need one of those pngs. It is used as a control to compare downloaded images against.)
- ```/var/lib/gems/2.5.0/gems/himawari-0.1.1/lib/himawari/process.rb:14:in `size': No such file or directory @ rb_file_s_size - /var/lib/gems/2.5.0/gems/himawari-0.1.1/lib/himawari/no_image.png (Errno::ENOENT)```

# 0.1.1
- Updated Readme
- Added changelog
- Set a minimum Ruby version to >=2.5
- Removed *.png files from the gem to reduce its size

# 0.1.0
- Initial release


spec/test.rb
require 'minitest/autorun'
require 'himawari'

class HimawariTest < Minitest::Test
  def test_english_hello
    assert_equal 'hello world', Himawari.get_pic
  end
end


tests/test.rb
require 'minitest/autorun'
require 'himawari' 

class HimawariTest < Minitest::Test
  def test_english_hello
    assert_equal "hello world",
      Hola.hi("english")
  end

  def test_spanish_hello
    assert_equal "hola mundo",
      Hola.hi("spanish")
  end
end


spec/test_net.rb
# frozen-string-literal: true

require 'test_helper'

class TestNet < HimawariTest
  # Do the main methods function w/o crashing?
  def test_get_pic
    puts "Try to get a pic from #{@timestamp}"
    assert_equal true, Himawari.get_pic(datetime: @timestamp, workdir: @@workdir)
  end

  def test_get_pic_blacklisted_wifi
    current_wifi = case Himawari::OsUtils.os
                   when :mac
                     `networksetup -getairportnetwork en0`
                   when :linux
                     `iwgetid -r`
                   end

    capture_stdout do
      assert_equal false, Himawari.get_pic(datetime: @timestamp, workdir: @@workdir, blacklist: [current_wifi])
    end
  end

  def test_get_pics
    puts "Try to get pics in #{@timestamp} ~ #{@timestamp + 3600 * 2}"
    assert_equal true, Himawari.get_pics(from: @timestamp, to: @timestamp + 3600 * 2, workdir: @@workdir)
  end
end


lib/himawari.rb
# frozen-string-literal: true

require_relative './himawari/base'
require_relative './himawari/os_utils'
require_relative './himawari/net_utils'
require_relative './himawari/process'
require_relative './himawari/download'

# encapsulates all the functions pertaining to acquiring images from the himawari8 satellite in near real-time
# from HIMAWARI_URL (defined in the Base Class)
module Himawari
  # all-in-one method. downloads images, sets backgrounds, crontabs, whatever
  # @param params [Hash] any combination of the acceptable command line args you can throw at it.
  #    Plz see README for the list/description
  # @return nothing useful
  def self.autorun(params = {})
    h = Download.new(params)
    h.cron_action ? h.crontab : h.update_backgrnd ^ h.start
  end

  # downloads 1 picture from himawari website, closest (but in the past of) the `:datetime` provided in the params
  # @param [Hash] any combination of the acceptable command line args + THE REQUIRED `:datetime` additional stamp
  # @return [true, false] on success/failure
  def self.get_pic(params = {})
    t = validate(params[:datetime])
    Download.new(params).pic(t, OsUtils.tenmin(t.min, t.min))
  end

  # downloads many pictures from himawari website, in the range of [:from, :to] provided in the params
  # @param [Hash] any combination of the acceptable command line args + THE REQUIRED `:from` && `:to` timestamps
  # @return [true, false] on success/failure
  def self.get_pics(params = {})
    Download.new(params).pics(validate(params[:from]), validate(params[:to]))
  end

  # validates the user-provided timestamps coming from `get_pic` or `get_pics` methods
  # @param stamp [DateTime, String]
  # @return [DateTime] of the stamp, or the most recent round 10-minute mark to Time.now - 10.minutes
  def self.validate(stamp)
    return stamp if stamp.is_a? Time
    return Time.parse("#{stamp}+00:00") if stamp.is_a? String

    t = Time.now.utc - 600 # 600secs == 10.minutes ago
    Time.new(t.year, t.month, t.day, t.hour, t.min / 10 * 10, 0, '+00:00')
  end
end


lib/himawari/base.rb
# frozen-string-literal: true

require 'httparty'
require 'fileutils'
require 'pathname'
require 'optparse'
# require 'pry'

module Himawari
  # This Aspect Ratio is used to try and download the most appropriate Tiles from Himawari,
  # while omitting those that won't be visible. (used by the :focus param)
  MONITOR_ASPECT = 16.0 / 9
  # int; update the pic once every `UPDATE_RATE` minutes
  UPDATE_RATE = 2
  # Yep, that's the big one. Where we actually download everything from
  HIMAWARI_URL = 'https://himawari8-dl.nict.go.jp/himawari8/img/D531106'

  # Local (JST) midnight happens @14:10 UTC, so we have to use pics in interval of [2.days.ago@14:40 .. 1.day.ago@14:40]
  LOCAL_MIDNIGHT = '1410'

  # sets up the methods for Himawari that do not have to do with accessing the internet. ie:
  #   - it sets up all the default and customized params, builds awareness of the locally stored pics
  #   - implements "setting" the background (copies a downloaded image into a different folder)
  class Base
    attr_reader :now, :app_root, :work_path, :data_path, :destination_path, :focus, :mode, :resolution,
                :latest_local, :latest_remote, :verbose, :by_schedule, :blacklist_wifi, :cron_action

    def initialize(params = {})
      init_paths(params[:workdir])
      @destination_path = params[:destination]

      @now = Time.now.utc
      @focus = params[:focus] || :top
      @mode = params[:mode] || :day
      @resolution = params[:resolution] || 2
      @latest_local = find_latest_local
      @latest_remote = nil
      @verbose = params[:verbose]
      @by_schedule = params[:by_schedule]
      @blacklist_wifi = params[:blacklist] || []
      @cron_action = params[:cron]
    end

    # @return [true, false] whether our local pics are up to date w/ himawari website
    def up_to_date?
      puts "Latest local:    #{latest_local[:timestamp]}" if verbose
      if now - latest_local[:timestamp] < 10 * 60
        puts 'Local pic is up to date; no need to go online.' if verbose
        return true
      end
      false
    end

    # sets sleeps for a bit, and then copies a selected picture to the `destination_path`
    # from where the OS can use it to set a background pic
    # @param img [String] full path to the picture
    def background(img)
      return false unless img && params_valid?

      cmd = "sleep 5 ; rm -f #{destination_path}/*.png ; cp #{img} #{destination_path}"
      # "osascript -e 'tell application \"System Events\" to tell DESKTOP to set picture to \"#{img}\"'"
      puts cmd if verbose
      system(cmd)
    end

    # useful for the "cycle through our downloaded pics and pic one for background" mode
    # checks that we have enough pictures for a whole :day of rotation, and calls the `background` method
    # with the appropriate picture's path based on current timestamp
    def update_backgrnd
      return false unless mode == :day && params_valid?

      # only rotate the background if we have one complete rotation of the globe saved
      sexy_pics = full_24hrs_available
      # puts sexy_pics
      return false unless sexy_pics

      i = (now.hour * 60 + now.min) / UPDATE_RATE % sexy_pics.count
      puts "#{i} :: #{now.hour * 60 + now.min} % #{sexy_pics.count} switching to #{sexy_pics[i]}" if verbose
      background(sexy_pics[i])
    end

    # public method exposing setting/clearing the crontab
    # @param action [Symbol] should we set the cron, or clear it?
    def crontab(action = nil)
      @cron_action = action if action
      return false unless cron_action && params_valid?

      cmd = '* * * * * PATH=/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin himawari ' \
            "-s -m #{mode} -f #{focus} -r #{resolution} -d '#{destination_path}' -w '#{work_path}' -b #{blacklist_wifi.join(',')}"
      OsUtils.crontab(cmd, cron_action)
      cmd
    end

    # verifies the required params that user might have messed up.
    # @return [true, false] depending on the validity of the parameters supplied
    def params_valid?
      unless %i[full top mid low].include?(focus) && %i[live day].include?(mode) &&
             [2, 4, 8, 16, 20].include?(resolution) && blacklist_wifi.is_a?(Array) &&
             (!destination_path || (destination_path && File.directory?(destination_path) && destination_path != '/'))
        puts 'Invalid params. Please double check them.'
        return false
      end
      true
    end

    private

    # @return [Array <String>] of pic_names to choose a `background` from, or
    #    false if we do not have enough pictures for a full 24hr rotation
    def full_24hrs_available
      prev_pic = ''
      sexy_pics = []

      Dir["#{data_path}/h_*.png"].sort { |x, y| y <=> x }.each do |pic|
        pic_time = pic[-8, 4]
        return sexy_pics if sexy_pics.count.positive? && pic_time < LOCAL_MIDNIGHT && LOCAL_MIDNIGHT <= prev_pic

        sexy_pics.unshift(pic) if sexy_pics.count.positive? || pic_time < LOCAL_MIDNIGHT && LOCAL_MIDNIGHT <= prev_pic
        prev_pic = pic_time
      end
      false
    end

    # relative to the location of the script: Pathname.new(File.expand_path('../', __FILE__))
    # relative to the current working dir:    Dir.pwd
    # sets the paths for the script
    # @param workdir [String] is taken from the param that the user supplied (or defaults to pwd)
    # (it's where we want to save the images from himawari)
    # the directory will be created if it doesn't exist yet
    def init_paths(workdir)
      @app_root = workdir && File.directory?(workdir) ? workdir : Dir.pwd
      @work_path = @app_root
      @data_path = "#{@app_root}/data"

      @app_root = Pathname.new(File.expand_path(__dir__))
      Dir.mkdir(@data_path) unless File.exist?(@data_path)
    end

    # scans the `workdir` and
    # @return [DateTime] of the most recent picture saved in `workdir`
    def find_latest_local
      # we can't select by timestamp, because those could be messed up. Instead, just search for "latest" by filename
      # latest_pic = `ls -t #{data_path}/h_* | head -1`
      twodays_ago = now - 86_400 * 2
      latest = { timestamp: nil }
      failsafe = { filename: nil, timestamp: Time.new(twodays_ago.year, twodays_ago.month, twodays_ago.day, 0, 0, 0, '+00:00') }

      Dir["#{data_path}/h_*.png"].each do |pic|
        stamp = Time.parse("#{pic[0..-5].insert(-3, ':')}:00+00:00")
        `touch -t #{stamp.strftime('%Y%m%d%H%M.%S')} #{pic}`
        latest = { filename: pic, timestamp: stamp } if !latest[:timestamp] || latest[:timestamp] < stamp
      end

      latest[:timestamp] && latest[:timestamp] > twodays_ago ? latest : failsafe
    end
  end
end


lib/himawari/process.rb
# frozen-string-literal: true

module Himawari
  # these methods provide checking and (trying to) fix the Tiles that we downloaded from the web.
  # sometimes, the connection might be bad, and instead of an actual PNG for a Tile, we get a blank empty file.
  # in other times, we get a special "no_image" png. This one is annoying. It IS a real picture,
  # but it most CERTAINLY is NOT a picture of the Earth. (These normally come back during 0240 and 1440 timestamps)
  # -- Himawari does not take photos when the Sun is in the view of the Satellite. Well, we wouldn't be able to see
  # the Earth at those times anyway due to the Sun's brightness.
  module Process
    private

    # Checks all the tiles downloaded for valid content.
    # Aims to throw out either empty tiles, or the weird "no-image" pngs
    # @return true on success
    def check_tiles
      @control_size ||= File.size("#{app_root}/no_image.png")
      bad_tiles = {}
      Dir["#{data_path}/t_*.png"].each do |tile|
        bad_tiles = process_bad_tiles(bad_tiles, tile) if bad_tile?(tile)
      end
      recover_bad_sectors(bad_tiles)
    end

    # Called from the loop in `check_tiles`. Does the dirty work of actually checking each tile
    # @param bad_tiles [Array <String>] the array that gets supplemented w/ any new tiles
    # @param tile [String] the yet-unknown-to-be-good tile image we just downloaded from the net
    # @return [Array <String>] of bad tile filenames
    # bad_tiles structure:
    # { "t_2019-11-11T0240": [ [ [tile_url_1, tile_url_2, tile_url_3] ], [array_of_tiles], [array_of_tiles] ] }
    def process_bad_tiles(bad_tiles, tile)
      i = File.basename(tile[0..-9])
      ending = tile[-7, 3]
      stamp = Time.parse("#{i.dup.insert(-3, ':')}:00+00:00")
      bad_tiles[i] = [] unless bad_tiles[i]
      bad_tiles[i] << [
        himawari_format(stamp, ending),
        himawari_format(stamp - 600, ending),
        himawari_format(stamp - 1_200, ending)
      ]
      # `cp #{tile} #{tile.gsub('/t_', '/x_')}`
      # no need to delete individual "bad tiles" because we will try to recover them.
      # But upon failure, will delete the whole himawari image
      # File.delete(tile)
      bad_tiles
    end

    # @param stamp [DateTime]
    # @param tile_end [String] the tail of the tile's filename. its extension?
    # @return [String] the format of how the images are named on the himawari website
    def himawari_format(stamp, tile_end)
      "#{stamp.year}/#{stamp.month}/#{stamp.day}/#{stamp.hour}#{stamp.min}00_#{tile_end}"
    end

    # verifies `file` to be an actual png
    # @param file [String]
    # @return true if the file provided is a PNG; false otherwise
    def png?(file)
      File.size(file).positive? && IO.read(file, 4).force_encoding('utf-8') == "\x89PNG"
    end

    # verifies the `tile` to be a good png and not a "no_image.png"
    # @param tile [String] filename
    # @return true if the tile is ok; false otherwise
    def bad_tile?(tile)
      !png?(tile) || File.size(tile) == @control_size && system("cmp #{tile} #{app_root}/no_image.png")
    end

    # attempts to recover the tiles in the `bad_tiles` array by downloading them again
    # or downloading a slightly older tile in the same position as the problem tile
    # @param bad_tiles [Array]
    # @return a message to the user of the conditional success of the tiles' recovery
    #    usually it deletes the whole image upon recovery failure... :(
    def recover_bad_sectors(bad_tiles)
      bad_tiles.each do |bad_pic_name, sectors|
        pic_not_good = true

        if sectors.count >= resolution # don't try to fix/recover pics that have too many pieces missing
          puts "#{bad_pic_name} has #{sectors.count} pieces missing. Too many. Can't recover & will delete the whole thing!"
        else
          puts "#{bad_pic_name} has #{sectors.count} pieces missing. Will try to recover/fill in with older data."
          p sectors
          tile = "#{data_path}/#{bad_pic_name}-#{sectors[0][-3, 3]}.png"
          sectors.each do |bad_tile|
            `curl -sC - "#{HIMAWARI_URL}/#{resolution}d/550/#{bad_tile}.png" > #{tile}`
            if bad_tile?(tile) # yep, it's bad...
              # File.delete(tile)
              pic_not_good = true
            else
              puts "#{tile} was recovered using #{bad_tile}!"
              pic_not_good = false
              break
            end
          end
        end
        # either too many pieces missing, or couldn't recover the wholes. It will look UGLY after reassembly,
        # so just bite the bullet and get rid of the whole thing
        `rm #{data_path}/#{bad_pic_name}*` if pic_not_good
      end
    end
  end
end


lib/himawari/net_utils.rb
# frozen-string-literal: true

module Himawari
  # Network-related methods to help establish whether there is network access and
  # figure out if we are on a possibly "metered" connection (via a hard-coded SSID blacklist)
  module NetUtils
    # Checks our connection and as a bonus...
    # @return [DateTime] timestamp of the latest himawari image; also as a side-efffect:
    #    true, false whether we even have connection to the website
    def internet_connection?
      uid = now.to_i * 1_000
      r = HTTParty.get("#{HIMAWARI_URL}/latest.json?uid=#{uid}", { timeout: 10 })
      # puts "#{HIMAWARI_URL}/latest.json?uid=#{uid}" if verbose
      if r.code == 200
        puts "Latest Himawari: #{r['date']}" if verbose
        @latest_remote = Time.parse("#{r['date']}+00:00")
        return true
      end
      false
    end

    private

    # Sometimes we don't want to use a particular network to download large amounts of data...
    # @return [true, false] whether the current (wifi) network is blacklisted in the blacklist_wifi [Array]
    def blacklisted_wifi?
      current_wifi = case OsUtils.os
                     when :mac
                       `networksetup -getairportnetwork en0`
                     when :linux
                       `iwgetid -r`
                     end
      puts current_wifi if verbose
      current_wifi && blacklist_wifi.any? { |wifi| current_wifi.include?(wifi) }
    end
  end
end


lib/himawari/os_utils.rb
# frozen-string-literal: true

module Himawari
  # all the misc. functions for dealing with CLI and/or OS...
  module OsUtils
    # we gotta know what the user's OS is! (because we rely on command line utils to process images)
    # @return [Symbol] representing the OS being used
    def self.os
      case RUBY_PLATFORM
      when /win32/
        :win
      when /linux/
        :linux
      when /darwin/
        :mac
      when /freebsd/
        :freebsd
      else
        :unknown
      end
    end

    # rubocop:disable Metrics/MethodLength, Metrics/BlockLength
    # This is just a list of args that the CLI can accept
    # @return [Hash] of cleaned/verified params
    def self.parse_cli_args
      params = {}

      OptionParser.new do |opts|
        opts.banner = 'Usage: himawari [params]'

        opts.on('-f', '--focus STRING', String, 'Which section of the planet to focus on? ' \
                                                'Valid values are `full`, `top`, `mid`, `low`. Default is `top`.') do |o|
          params[:focus] = o.to_sym if %w[full top mid low].include? o
        end

        opts.on('-m', '--mode STRING', String, 'Valid values are `day` (cycles pics in the `destination` folder from the' \
                                               'most recent day) or `live` (copies the latest photo downloaded ' \
                                               'to `destination`) Default is `day`.') do |o|
          params[:mode] = :live if o == 'live'
        end

        opts.on('-r', '--resolution INT', Integer, 'Adjust the resolution of the downloaded image. Valid numbers are ' \
                                                   '2, 4, 8, 16, 20. 20 is the highest resolution and 2 is the default. ' \
                                                   'For a 4k-monitor a setting of 4 seems sufficient.') do |o|
          params[:resolution] = o if o <= 20 && o.positive? && o.even?
        end

        opts.on('-d', '--destination PATH', String, 'The folder where to copy a background image. If left blank, images will ' \
                                                    'just be downloaded, but won\'t be copied anywhere afterward.') do |o|
          params[:destination] = o if File.directory?(o)
        end

        opts.on('-w', '--workdir PATH', String, 'The folder where to save all the downloaded pics. If left blank, ' \
                                                'images will be saved to the `./data` directory relative to your ' \
                                                'current path of working dir.') do |o|
          params[:workdir] = o if File.directory?(o)
        end

        opts.on('-b', '--blacklist STRING,STRING...', Array, 'Blacklist SSIDs of networks from which we do not want to go ' \
                                                             'online to download new images.') do |o|
          params[:blacklist] = o
        end

        opts.on('-c', '--cron STRING', String, 'Can `set`/`clear` cron with the specified params, so we can update the images ' \
                                               'automatically') do |o|
          params[:cron] = o.to_sym if %w[set clear].include? o
        end

        opts.on('-v', '--verbose', 'Increase verbosity: mostly for debugging') do |o|
          params[:verbose] = o
        end

        opts.on('-s', '--schedule', 'Flag for determining when the script is run by schedule/automatically.') do |o|
          params[:by_schedule] = o
        end

        opts.on('-h', '--help', 'Prints this help & exits') do
          puts opts
          exit
        end
      end.parse! # (into: params)

      params
    end
    # rubocop:enable Metrics/MethodLength, Metrics/BlockLength

    # Ok, this is a weird method. On Mac, `script` just ran normally through ruby. But, on linux...
    # It is refusing to interpolate variables within the `script` string properly.
    # So, we'll make a temporary bash script, execute it, and then delete it.
    # @param script [String] the temp script's filename
    # @param command [String] what we actually want to be done... (probably process pics in parallel)
    # @return nothing useful (true on success)
    def self.scriptify_sys(script, command)
      `echo "#!/bin/bash\n#{command}" > #{script}`
      `chmod +x #{script} && #{script}`
      `rm #{script}`
    end

    # tiny helper method. converts normal minutes into the closest (floored) round ten-minutes.
    # (ie like 10, 20, 30, 40, 50). Himawari website uses these in the tile filenames...
    # @param from [Integer]
    # @param to [Integer] these should be pretty self-explanatory. should be positive and w/in 0-59 range.
    # @return [String] the formatted minute range string for `parallel` to use to download this range of pics
    def self.tenmin(from = 0, to = 59)
      "{#{from / 10}..#{to / 10}}"
    end

    # to force crossfade:
    # https://apple.stackexchange.com/questions/141834/applescript-to-change-desktop-image-on-all-monitors

    # set picture rotation to 1 -- turn on wallpaper cycling
    # set change interval to -1 -- force a change to happen right now
    # delay 1.5 -- wait a bit to allow for the fade transition - you may want to play w/ this #
    # set picture of item N of theDesktops to POSIX file ("/Users/vladimir/chie/lib/space/data/h_2019-11-06T0220.png")
    #   -- set wallpaper to wallpaper you want
    # set picture rotation to 0  -- turn off wallpaper cycling

    # tell application "System Events"
    #     tell every desktop
    #     tell desktop 1
    #         set pictures folder to "/Library/Desktop Pictures"
    #         set picture rotation to 2 -- using interval
    #         set change interval to 1800
    #         set random order to true
    #     end tell
    #     tell desktop 2
    #         set pictures folder to "/Library/Desktop Pictures/Mine"
    #         set picture rotation to 2 -- using interval
    #         set change interval to 1800
    #         set random order to true
    #     end tell
    # end tell

    # to silence the status mails, add MAILTO='' at the top of the crontab manually
    # Sets or clears the crontab!
    # @param cmd [String] the himawari script + all the passed args for it
    # @param action [Symbol] what to do? :set the cron? or :clear the cron?
    # @return nothing useful
    def self.crontab(cmd, action)
      if action == :set
        `(crontab -l ; echo \"#{cmd}\") 2>&1 | grep -v \"no crontab\" | sort | uniq | crontab -`
      else
        `(crontab -l ; echo \"#{cmd}\") 2>&1 | grep -v \"no crontab\" | grep -v himawari | sort | uniq | crontab -`
      end
    end
  end
end
